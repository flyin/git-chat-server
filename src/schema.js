const fs = require('fs');
const { join } = require('path');

const { merge, flatten } = require('lodash');
const { makeExecutableSchema } = require('graphql-tools');

const chatResolvers = require('./resolvers/chat');
const settingsResolvers = require('./resolvers/settings');
const userResolvers = require('./resolvers/user');


const rootSchema = [`
type Query {
  getSettings: Settings!
  getChannels: [Channel]
  getChannel(channelId: ID!): Channel!
  getMessages(channelId: ID!): [Message]
  getUser(userId: ID): User
}

type Mutation {
  createChannel(name: String!): Channel
  createMessage(text: String!, channelId: String!): Message!
  createUser(email: String! password: String!): User
  createTokenByCode(code: String!): Token
  createTokenByPassword(email: String! password: String!): Token
}

type Subscription {
  onMessage(channelId: ID!): Message
  onChannel: Channel!
}
`];

const schemasPath = join(__dirname, 'schemas');

module.exports = makeExecutableSchema({
  resolvers: merge(chatResolvers, settingsResolvers, userResolvers),

  typeDefs: [
    ...rootSchema,
    ...flatten(fs.readdirSync(schemasPath).map(file => require(join(schemasPath, file))))
  ]
});
