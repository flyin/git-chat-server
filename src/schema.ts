import { merge, concat } from 'lodash';
import { makeExecutableSchema } from 'graphql-tools';
import { chatResolvers, settingsResolvers, userResolvers } from 'resolvers';
import { chatSchema, settingsSchema, userSchema } from 'schemas';

const rootSchema = [`
type Query {
  getSettings: Settings!
  getChannels: [Channel]
  getChannel(channelId: ID!): Channel!
  getMessages(channelId: ID!): [Message]
  getUser(userId: ID): User
}

type Mutation {
  createChannel(name: String!): Channel
  createMessage(text: String!, channelId: String!): Message!
  createUser(email: String! password: String!): User
  createTokenByCode(code: String!): Token
  createTokenByPassword(email: String! password: String!): Token
}

type Subscription {
  onMessage(channelId: ID!): Message
  onChannel: Channel!
}
`];

export default makeExecutableSchema({
  resolvers: merge(chatResolvers, settingsResolvers, userResolvers),
  typeDefs: concat([], rootSchema, chatSchema, settingsSchema, userSchema)
});
