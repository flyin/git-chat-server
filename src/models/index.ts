export { default as Channel } from './channel';
export { default as Message } from './message';
export { default as User } from './user';
