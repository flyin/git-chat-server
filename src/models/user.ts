import * as mongoose from 'mongoose';
import * as validator from 'validator';
import * as bcrypt from 'bcrypt';

const Schema = mongoose.Schema;
const SALT_FACTOR = 5;

export interface UserModel extends mongoose.Document {
  avatar: string
  email: string
  github?: GithubModel
  isAdmin: boolean
  password: string
  passwordIsValid: (password: string) => Promise<boolean>
}

export interface GithubModel extends mongoose.Document {
  accessToken: string
  githubId: number
  name: string
  refreshToken: string
  scopes: [string]
}

const Github = new Schema({
  accessToken: { required: true, type: String },
  githubId: { required: true, type: Number },
  name: { default: null, type: String },
  refreshToken: { default: null, type: String },
  scopes: { type: [String] },
});

const User = new Schema({
  avatar: { default: null, type: String },

  email: {
    required: true,
    trim: true,
    type: String,
    unique: true,
    validate: [validator.isEmail, 'Email address `{VALUE}` is incorrect']
  },

  github: { required: false, type: Github },
  isAdmin: { default: false, type: Boolean },
  password: { default: null, type: String },
}, {
  timestamps: true
});

User.pre('save', async function (this: UserModel, next) {
  if (!this.isModified('password')  && this.password) {
    return next();
  }

  try {
    const
      salt = await bcrypt.genSalt(SALT_FACTOR);

    this.password = await bcrypt.hash(this.password, salt);
    next()
  } catch (err) {
    next(err)
  }
});

User.methods.passwordIsValid = async function (this: UserModel, password: string) {
  return await bcrypt.compare(password, this.password)
};

module.exports = mongoose.model<UserModel>('User', User);
