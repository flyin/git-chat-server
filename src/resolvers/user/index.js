module.exports = {
  Mutation: {
    createTokenByCode: require('./create-token-by-code'),
    createTokenByPassword: require('./create-token-by-password'),
    createUser: require('./create-user')
  },

  Query: {
    getUser: require('./get-user')
  }
};
