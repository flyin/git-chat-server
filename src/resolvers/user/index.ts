import createUser from './create-user';
import createTokenByPassword from './create-token-by-password';
import createTokenByCode from './create-token-by-code';
import getUser from './get-user';

export default {
  Mutation: {
    createTokenByCode,
    createTokenByPassword,
    createUser
  },

  Query: {
    getUser
  }
};
