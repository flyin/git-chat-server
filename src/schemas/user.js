module.exports = [`
type Github {
  _id: ID!
  githubId: ID!
  name: String,
  scopes: String
}

type User {
  _id: ID!
  avatar: String
  isAdmin: Boolean
  email: String!
  github: Github
}

type Token {
  _id: ID!
  token: String!
  iat: String!
  exp: String!
  user: User!
}
`];
