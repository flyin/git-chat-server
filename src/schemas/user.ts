export default [`
type Github {
  _id: ID!
  avatar: String
  githubId: ID!
  name: String,
  scopes: String
}

type User {
  _id: ID!
  isAdmin: Boolean
  email: String!
  github: Github
}

type Token {
  _id: ID!
  token: String!
  iat: String!
  exp: String!
  user: User!
}
`];
