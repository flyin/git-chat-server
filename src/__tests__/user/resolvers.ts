import { get } from 'lodash';
import { graphql } from 'graphql';

import { initTestCase } from 'utils/tests-utils';
import schema from 'schema';
import userResolvers from 'resolvers/user';
import { mongoose } from 'services/mongoose';
import { UserModel } from 'models/user';

initTestCase();

const { createUser } = userResolvers.Mutation;
const userModel = mongoose.model<UserModel>('User');

describe('user resolvers', () => {
  describe('mutation', () => {
    const userCreateQuery = `
        mutation user($email: String!, $password: String!) {
          createUser(email: $email, password: $password) {
            email
            isAdmin
          }
        }
      `;

    it('user can register with email and password', async () => {
      const user = { email: 'a@b.ru', password: '123' };
      const result = await graphql(schema, userCreateQuery, {}, {}, user);
      expect(result.data).toMatchSnapshot();
      expect(result.errors).toBeUndefined();
    });

    it('user cant register with invalid data', async () => {
      const user = { email: 'a@', password: '123' };
      const result = await graphql(schema, userCreateQuery, {}, {}, user);
      expect(result.data).toBeNull();
      expect(result.errors).toMatchSnapshot();
    });

    it('user can exchange email and password for token', async () => {
      const user = { email: 'aa@aa.ru', password: '123' };
      await createUser(null, user);

      const query = `
        mutation token($email: String!, $password: String!) {
          createTokenByPassword(email: $email, password: $password) {
            token
            iat
            exp
            user {
              email
            }
          }
        }
      `;

      const result = await graphql(schema, query, {}, {}, user);
      const returnData = get(result.data, 'createTokenByPassword', {});

      expect(result.errors).toBeUndefined();
      expect(returnData).toHaveProperty('token');
      expect(returnData).toHaveProperty('user');
      expect(get(returnData, 'user')).toMatchObject({ email: user.email });
    });

    it('user can register with github token');
    it('user can reset their password');
    it('user can connect github to account');
    it('user can disconnect github from account');
  });

  describe('query', () => {
    const query = `
        query user($userId: ID) {
          getUser(userId: $userId) {
            email
            isAdmin
          }
        }
      `;

    it('user can access to own profile', async () => {
      let result;
      const user = await createUser(null, { email: 'aa@aa.ru', password: '123' });
      result = await graphql(schema, query, {}, { currentUser: Promise.resolve(user) });

      expect(result.errors).toBeUndefined();
      expect(get(result.data, 'getUser.email')).toBe(user.email);

      result = await graphql(schema, query, {}, { currentUser: Promise.resolve(user) }, { userId: 'some-exists-or-not-id' });

      expect(result.errors).toBeUndefined();
      expect(get(result.data, 'getUser.email')).toBe(user.email);
    });

    it('admin user can access to any account', async () => {
      let result;
      const user = await createUser(null, { email: 'aa@aa.ru', password: '123' });

      const adminUser = await userModel.create({
        email: 'admin@example.com',
        password: '123',
        isAdmin: true
      });

      result = await graphql(schema, query, {}, { currentUser: Promise.resolve(adminUser) }, { userId: user.id });

      expect(result.errors).toBeUndefined();
      expect(get(result.data, 'getUser.email')).toBe(user.email);

      result = await graphql(schema, query, {}, { currentUser: Promise.resolve(adminUser) }, { userId: adminUser.id });

      expect(result.errors).toBeUndefined();
      expect(get(result.data, 'getUser.email')).toBe(adminUser.email);
    });
  });

  describe('subscriptions', () => {
    it('notify socket then user is registered');
  });
});
