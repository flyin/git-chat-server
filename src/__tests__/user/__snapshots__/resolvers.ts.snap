// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user resolvers mutation user can register with email and password 1`] = `
Object {
  "createUser": Object {
    "email": "a@b.ru",
    "isAdmin": false,
  },
}
`;

exports[`user resolvers mutation user cant register with invalid data 1`] = `
Array [
  [GraphQLError: User validation failed: email: Email address \`a@\` is incorrect],
]
`;
